<a href="https://colab.research.google.com/github/data-better/ASL/blob/master/10%EA%B0%95_intro_to_rnn_keras.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
from keras.models import Sequential
from keras.layers import SimpleRNN, Dense, Embedding
from keras.preprocessing.text import one_hot
from keras.preprocessing.sequence import pad_sequences

## Parameter 개수 확인
### Simple RNN 1 layer

In [ ]:
model = Sequential()
model.add(SimpleRNN(32, input_shape=(3,10000), return_sequences=False))
model.add(Dense(1))
model.summary()

W0803 03:38:57.279271 140453068625792 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:74: The name tf.get_default_graph is deprecated. Please use tf.compat.v1.get_default_graph instead.

W0803 03:38:57.334946 140453068625792 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:517: The name tf.placeholder is deprecated. Please use tf.compat.v1.placeholder instead.

W0803 03:38:57.343529 140453068625792 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:4138: The name tf.random_uniform is deprecated. Please use tf.random.uniform instead.



_________________________________________________________________
Layer (type)                 Output Shape              Param #   
simple_rnn_1 (SimpleRNN)     (None, 32)                321056    
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 33        
Total params: 321,089
Trainable params: 321,089
Non-trainable params: 0
_________________________________________________________________


$$ h_t = \phi (x_t^T W_x + h_{t-1}^T W_h + b)$$
$W_x$: 10000개의 input 값과 32개 output($h_t$)과의 연결 = $32 \times 10000 = 320000$
$W_h$: 32개 $h_{t-1}$과 32개 $h_{t}$와의 연결 = $32 \times 32 = 1024$
$b$: 32개 bias

### Simple RNN with Embedding

In [ ]:
docs = ['Well done!',
		'Good work',
		'Great effort',
		'nice work',
		'Excellent!',
		'Weak',
		'Poor effort!',
		'not good',
		'poor work',
		'Could have done better.']
labels = np.array([1,1,1,1,1,0,0,0,0,0])
        
vocab_size = 50
encoded_docs = [one_hot(d, vocab_size) for d in docs]
print(encoded_docs)

max_length = 4
padded_docs = pad_sequences(encoded_docs, maxlen = max_length, padding = 'post')
print(padded_docs)

[[32, 9], [12, 43], [6, 2], [42, 43], [15], [40], [28, 2], [19, 12], [28, 43], [38, 20, 9, 8]]
[[32  9  0  0]
 [12 43  0  0]
 [ 6  2  0  0]
 [42 43  0  0]
 [15  0  0  0]
 [40  0  0  0]
 [28  2  0  0]
 [19 12  0  0]
 [28 43  0  0]
 [38 20  9  8]]


In [ ]:
model = Sequential()
model.add(Embedding(vocab_size, 8, input_length = max_length))
model.add(SimpleRNN(32, return_sequences = False))
model.add(Dense(1))
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_5 (Embedding)      (None, 4, 8)              400       
_________________________________________________________________
simple_rnn_6 (SimpleRNN)     (None, 32)                1312      
_________________________________________________________________
dense_6 (Dense)              (None, 1)                 33        
Total params: 1,745
Trainable params: 1,745
Non-trainable params: 0
_________________________________________________________________


$$emb_{out} = W x$$
$$W: 50\times 8 = 400$$
$$h_t = \phi(x_t^T W_x + h_{t-1}^T W_h + b)$$
$W_x: 8 \times 32 = 256$: 8 dim의 input을 32개의 hidden으로 mapping

$W_h = 32 \times 32 = 1024$: 32차원의 hidden을 32차원으로..

$b: 32 \times 1$: 32개의 bias

### Simple RNN 2 layers

In [ ]:
from keras.models import Sequential
from keras.layers import SimpleRNN, Dense

model = Sequential()
model.add(SimpleRNN(32, input_shape=(3,10000), return_sequences=True))
model.add(SimpleRNN(32, input_shape=(3,10000), return_sequences=False))
model.add(Dense(1))
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
simple_rnn_2 (SimpleRNN)     (None, 3, 32)             321056    
_________________________________________________________________
simple_rnn_3 (SimpleRNN)     (None, 32)                2080      
_________________________________________________________________
dense_2 (Dense)              (None, 1)                 33        
Total params: 323,169
Trainable params: 323,169
Non-trainable params: 0
_________________________________________________________________


$$ h_t = \phi (x_t^T W_x + h_{t-1}^T W_h + b)$$
RNN layer 1

$W_x$: 10000개의 input 값과 32개 output($h_t$)과의 연결 = $32 \times 10000 = 320000$
$W_h$: 32개 $h_{t-1}$과 32개 $h_{t}$와의 연결 = $32 \times 32 = 1024$
$b$: 32개 bias

=> 320000 + 1024 + 32 = 321056

RNN layer 2

$W_x$: 32개의 input 값과 32개 output($h_t$)과의 연결 = $32 \times 32 = 1024$
$W_h$: 32개 $h_{t-1}$과 32개 $h_{t}$와의 연결 = $32 \times 32 = 1024$
$b$: 32개 bias

=> 1024 + 1024 + 32 = 2080

In [ ]:
from keras.models import Sequential
from keras.layers import LSTM, Dense

model = Sequential()
model.add(LSTM(32, input_shape=(3,10000), return_sequences=False))
model.add(Dense(1))
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
lstm_1 (LSTM)                (None, 32)                1284224   
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 33        
Total params: 1,284,257
Trainable params: 1,284,257
Non-trainable params: 0
_________________________________________________________________
